server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1911,82 +2244,246 @@ public boolean assertAllListenersResolved() {
-                    }
-                    shards.put(finishedShardId, updateSnapshotState.updatedState);
-                    executedTasks.add(updateSnapshotState);
-                    changedCount++;
-                } else if (executedTasks.contains(updateSnapshotState)) {
-                    // tasks that completed a shard might allow starting a new shard snapshot for the current snapshot
-                    final ShardSnapshotStatus existingStatus = entry.shards().get(finishedShardId);
-                    if (existingStatus == null || existingStatus.state() != ShardState.QUEUED) {
-                        continue;
+                if (updateSnapshotState.isClone()) {
+                    // The update applied to a shard clone operation
+                    final RepositoryShardId finishedShardId = updateSnapshotState.repoShardId;
+                    if (entry.snapshot().getSnapshotId().equals(updatedSnapshot.getSnapshotId())) {
+                        final ShardSnapshotStatus existing = entry.clones().get(finishedShardId);
+                        if (existing == null) {
+                            logger.warn("Received clone shard snapshot status update [{}] but this shard is not tracked in [{}]",
+                                    updateSnapshotState, entry);
+                            assert false : "This should never happen, master will not submit a state update for a non-existing clone";
+                            continue;
+                        }
+                        if (existing.state().completed()) {
+                            // No point in doing noop updates that might happen if data nodes resends shard status after a disconnect.
+                            iterator.remove();
+                            continue;
+                        }
+                        logger.trace("[{}] Updating shard clone [{}] with status [{}]", updatedSnapshot,
+                                finishedShardId, updateSnapshotState.updatedState.state());
+                        if (clones == null) {
+                            clones = ImmutableOpenMap.builder(entry.clones());
+                        }
+                        changedCount++;
+                        clones.put(finishedShardId, updateSnapshotState.updatedState);
+                        executedTasks.add(updateSnapshotState);
+                    } else if (executedTasks.contains(updateSnapshotState)) {
+                        // the update was already executed on the clone operation it applied to, now we check if it may be possible to
+                        // start a shard snapshot or clone operation on the current entry
+                        if (entry.source() == null) {
+                            // current entry is a snapshot operation so we must translate the repository shard id to a routing shard id
+                            final IndexMetadata indexMeta = currentState.metadata().index(finishedShardId.indexName());
```

Comment:
I've been a bit worried about re-resolving indices here as the cloned shard could reference an index from the snapshot that do not exist anymore, or could have been recreated with a different sharding, but after writing some tests I think it's OK as `existingStatus == null` later will prevent any difference in the number of shards.